File: Lecture 15: Binary Search Advanced Problems/aggressiveCows.cpp
#include<vector>
using namespace std;

bool isPossible(vector<int> &stalls, int k, int mid) {
    
    int cowCount = 1;
    int lastPos = stalls[0];
    
    for(int i=0; i<stalls.size(); i++ ){
        
        if(stalls[i]-lastPos >= mid){
            cowCount++;
            if(cowCount==k)
            {
                return true;
            }
            lastPos = stalls[i];
        }
    }
    return false;
}

int aggressiveCows(vector<int> &stalls, int k)
{
    sort(stalls.begin(), stalls.end());
    
   	int s = 0;
    int maxi = -1;
    for(int i=0; i<stalls.size(); i++) {
        maxi = max(maxi, stalls[i]);
    }
    int e = maxi;
    int ans = -1;
    int mid = s + (e-s)/2;
    
    while(s<=e) {
        if(isPossible(stalls, k, mid)) {
            ans = mid;
            s = mid + 1;
        }
        else
        {
            e = mid - 1;
        }
        mid = s + (e-s)/2;
    }
    return ans;
}

File: Lecture 15: Binary Search Advanced Problems/aggressiveCows.java
import java.util.*;

public class Solution 
{
    static boolean isPossible(ArrayList<Integer> stalls,int k,int mid){
        int cows = 1,n=stalls.size();
        int last = stalls.get(0);
        for(int i=0;i<n;i++){
            if(stalls.get(i)-last >= mid){
                cows++;
                if(cows==k)
                    return true;
                
                last = stalls.get(i) ;
            }
        }
        return false ;
    }
    public static int aggressiveCows(ArrayList<Integer> stalls, int k) 
    {
        int end = Collections.max(stalls);
        int start = 0,ans=-1;
        Collections.sort(stalls);
        
        int mid = start+(end-start)/2;
        while(start<=end){
            if(isPossible(stalls,k,mid)){
                ans = mid ;
                start = mid+1 ;
            }else{
                end = mid-1 ;
            }
            mid = start +(end-start)/2;
        }
        return ans;
    }
}